libbeat/common/kubernetes/watcher.go

```diff
@@ -2,260 +2,189 @@ package kubernetes
-type Client interface {
-	ListPods(ctx context.Context, namespace string, options ...k8s.Option) (*corev1.PodList, error)
-	WatchPods(ctx context.Context, namespace string, options ...k8s.Option) (*k8s.CoreV1PodWatcher, error)
-}
-
-// NewWatcher initializes the watcher client to provide a local state of
-// pods from the cluster (filtered to the given host)
-func NewWatcher(client Client, syncPeriod, cleanupTimeout time.Duration, host string) Watcher {
+// NewWatcher initializes the watcher client to provide a events handler for
+// resource from the cluster (filtered to the given node)
+func NewWatcher(client *k8s.Client, resource Resource, options WatchOptions) (Watcher, error) {
 	ctx, cancel := context.WithCancel(context.Background())
-	return &podWatcher{
+	w := &watcher{
 		client:              client,
-		cleanupTimeout:      cleanupTimeout,
-		syncPeriod:          syncPeriod,
-		nodeFilter:          k8s.QueryParam("fieldSelector", "spec.nodeName="+host),
+		options:             options,
 		lastResourceVersion: "0",
 		ctx:                 ctx,
 		stop:                cancel,
-		pods:                make(map[string]*Pod),
-		deleted:             make(map[string]time.Time),
-		bus:                 bus.New("kubernetes"),
 	}
+	switch resource.(type) {
+	case *Pod:
+		w.resourceList = &v1.PodList{}
+		w.k8sResourceFactory = func() k8s.Resource { return &v1.Pod{} }
+		w.resourceFactory = func() Resource { return &Pod{} }
+	case *Event:
+		w.resourceList = &v1.EventList{}
+		w.k8sResourceFactory = func() k8s.Resource { return &v1.Event{} }
+		w.resourceFactory = func() Resource { return &Event{} }
+	default:
+		return nil, fmt.Errorf("unsupported resource type for watching %T", resource)
+	}
+	return w, nil
+}
+
+func (w *watcher) AddEventHandler(h ResourceEventHandler) {
+	w.handler = h
+}
+
+func (w *watcher) buildOpts() []k8s.Option {
+	options := []k8s.Option{k8s.ResourceVersion(w.lastResourceVersion)}
+	if w.options.Node != "" {
+		options = append(options, filterByNode(w.options.Node))
+	}
+	return options
 }
 
-func (p *podWatcher) syncPods() error {
-	logp.Info("kubernetes: %s", "Performing a pod sync")
-	pods, err := p.client.ListPods(
-		p.ctx,
-		"",
-		p.nodeFilter,
-		k8s.ResourceVersion(p.lastResourceVersion))
+func (w *watcher) sync() error {
+	ctx, cancel := context.WithTimeout(w.ctx, w.options.SyncTimeout)
+	defer cancel()
+
+	logp.Info("kubernetes: Performing a resource sync for %T", w.resourceList)
 
+	err := w.client.List(ctx, w.options.Namespace, w.resourceList, w.buildOpts()...)
 	if err != nil {
+		logp.Err("kubernetes: Performing a resource sync err %s for %T", err.Error(), w.resourceList)
 		return err
 	}
 
-	p.Lock()
-	for _, apiPod := range pods.Items {
-		pod := GetPod(apiPod)
-		p.pods[pod.Metadata.UID] = pod
+	// ugly reflect since ResourceList does not exposed Items
+	// can be avoid by using k8s.io/client-go
+	list := reflect.ValueOf(w.resourceList).Elem().FieldByName("Items")
```

Comment:
You can avoid this if you create a new interface for resource list, and write adapters to get the Items from PodList and EventList, what do you think?

Reply:
Good idea! I will make a try. Thanks!